packages/world/src/utils/InventoryUtils.sol

```diff
@@ -40,46 +40,52 @@ struct ToolData {
   ObjectType toolType;
   uint16 slot;
   uint128 massLeft;
-  uint128 maxUseMass;
 }
 
 library InventoryUtils {
   function getToolData(EntityId owner, uint16 slot) public view returns (ToolData memory) {
     EntityId tool = InventorySlot._getEntityId(owner, slot);
     if (!tool.exists()) {
-      return ToolData(owner, tool, ObjectTypes.Null, slot, 0, 0);
+      return ToolData(owner, tool, ObjectTypes.Null, slot, 0);
     }
 
     ObjectType toolType = EntityObjectType._get(tool);
     require(toolType.isTool(), "Inventory item is not a tool");
 
-    uint128 maxMass = ObjectPhysics._getMass(toolType);
-    uint128 maxUsePerCall = maxMass / 10; // Limit to 10% of max mass per use
-
-    return ToolData(owner, tool, toolType, slot, Mass._getMass(tool), maxUsePerCall);
+    return ToolData(owner, tool, toolType, slot, Mass._getMass(tool));
   }
 
-  function useTool(EntityId owner, Vec3 ownerCoord, uint16 slot, uint128 useMassMax)
-    public
-    returns (ObjectType toolType)
-  {
+  function useTool(EntityId owner, uint16 slot, uint128 useMassMax) public returns (ObjectType toolType) {
     ToolData memory toolData = getToolData(owner, slot);
 
-    uint128 massReduction = toolData.getMassReduction(useMassMax);
-    applyMassReduction(toolData, ownerCoord, massReduction);
+    (, uint128 toolMassReduction) = toolData.getMassReduction(useMassMax, 1);
+    reduceMass(toolData, toolMassReduction);
     return toolData.toolType;
   }
 
-  function getMassReduction(ToolData memory toolData, uint128 useMassMax) internal pure returns (uint128) {
-    uint128 massReduction = useMassMax > toolData.maxUseMass ? toolData.maxUseMass : useMassMax;
-    if (toolData.massLeft <= massReduction) {
-      return toolData.massLeft;
+  function getMassReduction(ToolData memory toolData, uint128 massLeft, uint128 multiplier)
+    internal
+    view
+    returns (uint128, uint128)
+  {
+    if (toolData.toolType.isNull()) {
+      return (0, 0);
     }
 
-    return massReduction;
+    uint128 toolMass = ObjectPhysics._getMass(toolData.toolType);
+
+    // Limit to 10% of max mass per use
+    uint128 maxToolMassReduction = Math.min(toolMass / 10, toolData.massLeft);
+
+    uint128 massReduction = Math.min(maxToolMassReduction * multiplier, massLeft);
+
+    // Reverse operation to get the proportional tool mass reduction
+    uint128 toolMassReduction = massReduction / multiplier;
```

Comment:
hmm can there be rounding errors here? Or I guess it doesnt matter cuz this is just about picking how much mass to reduce from the tool

Reply:
yeah there will be rounding errors, but I believe it is fine because toolMassReduction will always be <= maxToolMassReduction